Refactor of statistics monitoring (global & process mem/cpu/network)

apingault opened the pull request:
BEGINRELEASENOTES
- Replace statistics (mem/cpu/network) implementation with implementation from [ROOT TSystem class](https://root.cern.ch/doc/master/classTSystem.html)
  - Better Handling of OS-dependent implementation
  - Had to copy/paste and adapt the implementation instead of just using it to tailor it to our needs.
  - Implements `cpuStats` + `networkStats` + partial implementation of `procStats` for `OSX` (missing cpu load)
  - Partial implementation of `procStats` for `Linux` (missing cpu load)
- Adapt/Add new exec to chec for theses stats:
  - Adapt executable to check for `memStats`
  - Add exec to check for `networkStats` and `cpuStats`
  - `procStats` are included in all these exec (i.e memory use from process is reported in `memStats`)   
  - Should I add an exec dedicated to check `procStats`?  
  - Windows implementation still missing
- Add `dqm_long` and `dqm_long_long` type

ENDRELEASENOTES

TODO: 
 - [x] Finish computation of CPU load from the process, both for osx and linux. 
 - [x] Finish linux implementation of NetStats (not computing rate for now) 
 - [ ] Add unit tests

Commit:
Replace memStats by impl. from ROOT TSystem.h

Better Handling of os-dependent implementation.
Had to copy/paste and adapt the implementation instead of just using it to tailor it to our needs.
Add implementation of CpuStats + NetStats and partial implementation of ProcessStats for OSX (missing cpu load)
Add partial  implementation of procStats for Linux (missing cpu load)
Add `dqm_long` and `dqm_long_long` type

Commit:
Adapt memstats exec to changes from last commit

Commit:
Add exec to test netStats and cpuStats

Commit:
Fix badly formated warning message

apingault commented:
TODO: 
 - [ ] Finish computation of CPU load from the process, both for osx and linux. 
 - [ ] Add unit tests

Commit:
Fix linux compilation

Commit:
Fix attribute name on linux

Commit:
Add Warning on Linux: network rate are not impl

Commit:
Fix linuxNetStats definition

Commit:
Fix wrong unit for rcv/snd data in linuxNetStats

apingault commented:
@rete Could you check locally that it compiles and run fine on linux?
I don't have a linux box properly configured to test it out right now

Commit:
Fix missing include on linux

Commit:
Fix bad include placement on linux

rete commented:
Mem stats are broken on linux : 
```shell
[2018-05-29 15:56:33.418] [main] [warning] [linuxProcStats] - Process cpu load has not been implemented yet!
Memory    Total (Mb)      Used (Mb)       Process (Mb)    Used (%)        Proc used (%)   
Resident  0               0               13.4            -nan            inf             
Virtual   23536           8227            77.4            35.0            0.3
```
Rates for net stats are not implemented by seeing the source code, which is also displayed when running `dqm4hep-netstats` : 

```shell
[2018-05-29 16:05:23.677] [main] [warning] [linuxNetStats] - Network rate has not been implemented yet!
                                                                    Total                                                               |	                                                Rate                                                        
Interface       Sent (kb)           Received (kb)       Sent packets        Rcvd packets        Rcvd errors         Sent errors         |	Sent (kb/s)         Rcvd (kb/s)         Sent (packets/s)    Rcvd (packets/s)    Sent (error/s)      Rcvd (error/s)      
enp0s25         1134806051          9890804578          4798848             10330207            0                   0                   |	0                   0                   0                   0                   0                   0                   
enp5s0          0                   0                   0                   0                   0                   0                   |	0                   0                   0                   0                   0                   0                   
lo              3192319             3192319             2537                2537                0                   0                   |	0                   0                   0                   0                   0                   0
```

Commit:
Fix broken rssMemory stat on linux

rete commented on source/src/Internal.cc:
```diff
@@ -28,229 +28,694 @@
-      stats.rssproc /= unit;
-      stats.vmproc = taskInfo.resident_size; // Display physical size, as virtual_size is meaningless in this context
-      stats.vmproc /= unit;
+      dqm_long_long free = {0LL};
```
Please change the name `free` here as it is reserved for memory

apingault replied:
Indeed, fixed now.

Commit:
Fix `free` var name, avoid mistaking it with free()

Commit:
Add process cpu load for OSX and Linux

apingault commented:
@rete Can you check the memory/cpu load are properly reported on linux now?

Commit:
Update cpuStats exec with process cpu load report

We now need processStats to exist outside of the while loop so we can keep track of the cpu consumption

Commit:
Fix typo

Commit:
Fix missing struct initialisation

Commit:
Add network rate for linux

Commit:
Fix typo for linux

Commit:
Fix linux memStat not updated

Commit:
Fix compilation if DQM4HEP_WITH_PROC_FS not defined

Commit:
Add some doc to static functions

rete commented:
Looks fine for me on linux now. Tell me when you are done

apingault commented:
Well I am just missing some unit test now. But I won't be able to implement that before next week at the earliest.
So if it's OK with you we can already merge that for now.

rete commented:
Quite difficult to implement unit tests here. It is OS dependent and the variables are defined at runtime. You can't guess anything. I'm happy with these changes and merge it.
